File: prog3/trab_java/utilitarios/capes/producoes/Traducao.java
package utilitarios.capes.producoes;

public class Traducao extends Producao implements Comparable<Traducao> {
    String editora;
    String idioma_traducao;

    public Traducao(String titulo, String natureza, String idioma, String cidade, int qtdPags, String editora, String idioma_traducao){
        super(titulo, natureza, idioma, cidade, qtdPags);
        this.editora = editora;
        this.idioma_traducao = idioma_traducao;
        this.tipo = 6;
    }

    @Override
    public int compareTo(Traducao o) {
        return this.natureza.compareTo(o.natureza) != 0 ?
        this.natureza.compareTo(o.natureza) : this.titulo.compareTo(o.titulo) != 0 ?
        this.titulo.compareTo(o.titulo) : this.idioma.compareTo(o.idioma) != 0 ?
        this.idioma.compareTo(o.idioma) : this.editora.compareTo(o.editora) != 0 ?
        this.editora.compareTo(o.editora) : this.cidade.compareTo(o.cidade) != 0 ?
        this.cidade.compareTo(o.cidade) : this.idioma_traducao.compareTo(o.idioma_traducao) != 0 ?
        this.idioma_traducao.compareTo(o.idioma_traducao) : this.qtdPags - o.qtdPags;
    }

    public static void imprimeCabecalho(){
        System.out.println("Natureza;Titulo;Idioma;Editora;Cidade;Traducao;Paginas");
    }

    @Override
    public String toString(){
        return (this.natureza + ";" + this.titulo + ";" + this.idioma + ";" + this.editora + ";" + this.cidade + ";" + this.idioma_traducao + ";" + this.qtdPags_string);
    } 
}

File: prog3/trab_cpp/Traducao.cpp
#include "Traducao.hpp"

Traducao::Traducao(string titulo, string natureza, string idioma, string cidade, int qtd_pags, string editora, string idioma_traducao): Producao(titulo, natureza, idioma, cidade, qtd_pags, 21){
    this->editora = editora;
    this->idioma_traducao = idioma_traducao;
}

Traducao::~Traducao(){}

/* Metodo que retorna a string que contem as informacoes de uma producao em formato csv. No caso, reorna uma string no seguinte formato:
 "<Natureza>;<Titulo>;<Idioma>;<Editora>;<Cidade>;<Traducao>;<Paginas>"*/
string Traducao::csv(){
    stringstream stream;
    stream << natureza << ";" << titulo << ";" << idioma << ";" << editora << ";" << cidade << ";" << idioma_traducao << ";" << _int_string(qtd_pags);
    return stream.str();
}

/* Metodo que retorna a string cabecalho. */
string Traducao::cabecalho(){
    return "Natureza;Titulo;Idioma;Editora;Cidade;Traducao;Paginas";
}

/* Operador de comparacao de uma producao com outra de mesmo subtipo. */
bool Traducao::operator < (Producao& o){
    Traducao& prod = (Traducao&)o;
    int cmp = natureza.compare(prod.natureza);
    if(cmp<0) return true; else if(cmp>0) return false;
    cmp = titulo.compare(prod.titulo);
    if(cmp<0) return true; else if(cmp>0) return false;
    cmp = idioma.compare(prod.idioma);
    if(cmp<0) return true; else if(cmp>0) return false;
    cmp = editora.compare(prod.editora);
    if(cmp<0) return true; else if(cmp>0) return false;
    cmp = cidade.compare(prod.cidade);
    if(cmp<0) return true; else if(cmp>0) return false;
    cmp = idioma_traducao.compare(prod.idioma_traducao);
    if(cmp<0) return true; else if(cmp>0) return false;
    cmp = qtd_pags - prod.qtd_pags;
    if(cmp<0) return true; else return false;
}